### DISCUSSION TITLE
Import two or more extracts.

### DISCUSSION THREAD

**Discussion by StyXman** [osm2pgsql-dev/osm2pgsql#2031] [Q&A] [answered]
Is it possible to import two or more extracts to the same db? For instance, just a bunch of countries, or maybe just two continents...

**Comment by pnorman** [comment]
You can use multiple schemas or table prefixes, but need to be aware of conflicts with the metadata table. Multiple imports to the same tables is not supported.

**Comment by mboeringa** [comment, marked as answer]
According to [this section of the osm2pgsql manual](https://osm2pgsql.org/doc/manual.html#working-with-multiple-input-files), osm2pgsql >= 1.4 should be able to read multiple input files of the same date if added to the command line, and automatically resolve conflicts.

**Reply by joto** [reply]
To be a bit more specific here: It doesn't resolve conflicts. You have to make sure there are none. If the input files don't fit together, everything will break and you get to keep the pieces.

**Reply by mboeringa** [reply]
@joto.

My reference to conflict resolving was just related to the manual stated functionality of not importing duplicate data, as I guess that is the most common type of possible "conflict" you might get with multiple extracts.

This does raise a question though, if osm2pgsql shouldn't have a command line parameter to force dropping of incomplete ways and relations during import.

As https://osm2pgsql.org/doc/manual.html#handling-of-incomplete-osm-data states, osm2pgsql will always try to a import even incomplete data. This could cause issues with multiple extracts in overlapping areas where geometries have been cut in extracts.

If osm2pgsql had an option to only import complete objects, the result of importing e.g. two neighbor countries, might be cleaner.

Maybe this could be a new option on the command line...?

**Reply by joto** [reply]
The problem is not duplicate data but inconsistent data. Duplicate data is fine, osm2pgsql will detect that. The problem is when extract A has version X of a way and extract B has version Y. Then you get into all sorts of problems. If you have extracts from two neighboring countries with some overlap and the extracts were generated from the same planet, everything will be fine.

And this has nothing to do with incomplete data, that's a whole other issue and not solvable in general. There is just no defined way how you can cut OSM data without looking at the specifics of that data. We are not adding more command line options to supposedly fix something which is not fixable. If you have a problem with cut data, the only way to solve that is to make your extract so large that you are "pretty sure" that everything you need is in it and cut after importing into the database. And what "pretty sure" is depends on what data you actually need and what you are doing with it.

**Comment by StyXman** [comment]
I guess that means it can also update from more than one diff file?

**Reply by joto** [reply]
In general: no. You usually need to "simplify" diff files, as described in the manual. And if you do that anyway, you might as well simplify them all together (using Osmium for instance) and then only use a single diff file.
